posts/2021-03-13-microprofile-config-2.0.adoc

```diff
@@ -0,0 +1,331 @@
+----
+
+[#ConfigValue]
+=== ConfigValue
+The new link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/org/eclipse/microprofile/config/ConfigValue.html[ConfigValue API class] allows you to retrieve details about a given config property into one convenient ConfigValue object. And it’s super easy to get hold of. All you have to do is inject the config property you’d like, as usual, only this time define the type as ConfigValue:
+
+[source,java]
+----
+@Inject
+@ConfigProperty(name="server.host")
+ConfigValue serverNameConfigValue;
+----
+
+With this, you can retrieve all the useful values with the get methods defined in the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/apidocs/[Javadoc]. For example, you can determine which ConfigSource was the “winning” one (the ConfigSource with the highest ordinal) for a config property defined in multiple locations by calling:
+
+[source,java]
+----
+serverNameConfigValue.getSourceName(); // returns “PropertiesConfigSource[source=file:/<path-to-file>/META-INF/microprofile-config.properties]”
+serverNameConfigValue.getSourceOrnial(); // returns “100”, the default ordinal value for META-INF/microprofile-config.properties
+----
+
+[#Property-Expression]
+=== Property Expression
+Property Expressions provide a way to set and expand variables in property values using the `${}` syntax. For example, the config property `server.url` defined in the <<example-ConfigSource, example ConfigSource>>  as `\http://${server.host}` will be resolved to `\http://example.org` since `server.host` is defined as `example.org`
+
+[source,java]
+----
+@Inject
+@ConfigProperty(name="server.url")
+String url; // retrieves the resolved value of “http://example.org”, or “http://example.test.org” if mp.config.profile is set to “testing”
+----
+
+You can also do some funky expressions, defining default values, composed expressions, and multiple expressions, which link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#property-expressions[the spec] covers really well.
+
+Previous working configurations might behave differently if the configuration in use contains values with Property Expressions syntax.
+
+[#Extra-Info]
+== Some extra info for the return journey
+
+For the following examples, we'll use a slightly more rogue example ConfigSource (let's call it "example ConfigSource v2"):
+[[example-ConfigSource2]]
+.META-INF/microprofile-config.properties
+[source]
+----
+empty.property=
+empty.array.prop=,
+ports=9080,9081,9082
+server.port=9080
+----
+
+[#Config-Value-Behaviour-Updates]
+=== Empty And Special Values Behaviour Updates
+The behaviour for "empty" and "special" config property values has been updated:
+
+* The easiest way to get your head around this is to look at the link:https://download.eclipse.org/microprofile/microprofile-config-2.0/microprofile-config-spec-2.0.html#_config_value_conversion_rules[conversion rule examples].
```

Comment:
A general formatting observation is that the font size for bullet pointed information is smaller than normal text ... getting a bit too small maybe?

Reply:
Hmm Yes. Also, when a bullet point is broken by a [Complex content](https://docs.asciidoctor.org/asciidoc/latest/syntax-quick-reference/#ex-complex) the proceeding information is back to normal (non bullet point) size.

Not sure who to speak to about this formatting. 

@jakub-pomykala and @austin0 would you have any idea please?